### ISSUE TITLE
Understanding the correct use of `wrap_cancel`

### ISSUE THREAD

**Issue by hellertime** [de-vri-es/async-shutdown-rs#7]
I have been thinking of using `async-shutdown` in a project of mine, and was experimenting with how its used.

To start, I have to be able to have a processing loop that may spawn a number of background tasks, all of which I want to cancel if a shutdown signal is sent.

My approach was to wrap each task in a `wrap_cancel` and send these into the background, like so (I'm using glommio for this example, but the issue should be independent of the runtime chosen):

```rust
fn main() {
    let shutdown_mgr = async_shutdown::ShutdownManager::<()>::new();
    let local_ex = glommio::LocalExecutor::default();

    local_ex.run(async move {
        loop {
            let _ = glommio::spawn_local(shutdown_mgr.wrap_cancel(async move {
                glommio::executor().yield_now().await;
                let mut m = std::collections::HashMap::new();
                for n in 1..10 {
                    m.insert("a", n);
                }
                m.remove("a");
            }))
            .detach();
        }
    });
}
```

What I noticed is that it seems the design of `async-shutdown` is to clone the waker of each of these tasks, and place it on a vector, which, on a shutdown signal is iterated so that each task managed can be shutdown.

This happens here: https://github.com/de-vri-es/async-shutdown-rs/blob/9ec5d4b31cbe5b4a226a6849b0f357437024ad26/src/shutdown_signal.rs#L44

This all seems well, and good, but it would seem that if the future wrapped by `wrap_cancel` exists, its waker is never released from this vector, meaning the vector grows unbounded if I use `wrap_cancel` as I have in this example.

My guess is this isn't the expected use of `wrap_cancel`, but I'm not sure what would be a better solution here?

**Comment by de-vri-es** [comment]
Hmm, good point. I think you mean *exits*, not *exists*, right? But it makes sense, when the future is dropped the waker should be removed.

I'll get on this right away :)

**Comment by de-vri-es** [comment]
Should be fixed with #8. Do you want to test it in your use case?

**Comment by de-vri-es** [comment]
Fix released with `v0.2.2`. Thanks for reporting the issue!

**Comment by de-vri-es** [comment]
Also backported it to 0.1 and released as `v0.1.4`, because according to crates.io most people are still on version 0.1 >.<
